### ISSUE TITLE
Combing LinesPolygonsCrossing to consider not-yet-printed polygons

### ISSUE THREAD

**Issue by mattwn** [Ultimaker/Cura#11330]
### EDIT: 

After further investigation I believe this is a bug, not a feature request. In short: Spiralize properly implements combing/avoid everywhere EXCEPT after a `;TYPE:SUPPORT` section ( or potentially just before a `;TYPE:WALL-OUTER`). My experiments and rationale are below.

### Is your feature request related to a problem?

I originally commented to [Issue 9969](https://github.com/Ultimaker/Cura/issues/9969), but I realized that issue was not 100% related, so I deleted the posts and moved to this new issue.

The problem is highlighted in red in the graphic below and the expected behavior is highlighted in cyan. The simple project to demo this behavior is also attached.

[CE3_VikinigHorn.zip](https://github.com/Ultimaker/Cura/files/7927027/CE3_VikinigHorn.zip)

This is Layer 17 in Cura Preview (so around layer 12 in G Code):

![image](https://user-images.githubusercontent.com/95894900/150821569-aad2f27e-8deb-4b1e-95bd-4053f30deb69.png)

In issue 9969, @nallath implied to the OP that the intended behavior of combing is not to consider future polygons. At least that was my interpretation, which may be incorrect. The quote was:

> The infill is printed first, so at that moment there is no other stuff on that layer to comb through [_edit:_ or avoid]. As such it will just take the fastest path.

By ignoring not-yet-printed polygons/features (especially internal holes and concave walls), stringers can be left where these voids will eventually be.

### Describe the solution you'd like

If combing considered not-yet-printed polygons/features on the same layer,  Any stringers/wisps would be in places that would eventually not matter and there would be an improvement in surface quality.

### Describe alternatives you've considered

We wrote a G Code post-processing script to replace the single straight line of the ;MESH-filename section with a few moves that go out and around, but it is a brute force time sink, not readily extensible to different parts, and interferes with nozzle pressure management.

### Affected users and/or printers

This will affect any user of combing, especially improving the surface quality of parts with internal holes or complex surfaces that are printed after infill.

### Additional information & file uploads

We lose many $100s per month on production parts to this failure mode. Mostly in production schedule slip due to reprinting failed parts.

**Comment by mattwn** [comment]
Upon further investigation, perhaps this is a bug when spiralize interacts with combing. When I disable `Spiralize` and Change Surface mode from Normal to `Surface`. The combing behavior works as expected and the slice follows the cyan line above, implying that combing is respecting note-yet-printed polygons. 

Interestingly, when `Surface Mode: Surface` and `Spiralize` are both set, the slice still combs as expected and the Z seam is visible up the entire part, implying that Surface supersedes/disables Spiralize.

In the end, we get a similar vase like part, just without the smooth z transition. If that is the price to avoid travels through the internal void, we'll take it. But it does imply a bug with the interaction of Spiralize and Combing. I admit, Spiralize would rarely interact with Combing for most users. For us, it happens on the first few mm of Z axis when we add supports to reinforce the connection of the vase to the build plate.

In short, even if a bug exists, it's probably not wide spread and this issue can be deferred or closed as desired.

**Comment by mattwn** [comment]
This would be a great case where the ability to change the slicer profile at a specific z height would be useful (beyond the limited Change at Z post processor).

Z = 0 to 20 mm: Slice with Profile I
Z = 20 to 100 mm: Slice with Profile J
Z = 100 mm to end: Slice with Profile K

**Comment by mattwn** [comment]
Another note, I attempted to use a cutting mesh on just the bottom portion of the bottle. With the global mesh being `Spiralize: True` and `Surface Mode:Normal` and the cutting mesh having `Spiralize:False` and `Surface Mode: Surface`. (I had to modify fdmprinter.def.json to enable adjusting spiralize per model).

It worked well EXCEPT, `Surface Mode: Surface` slightly expands the part resulting in a visible step change of diameter. Typically I might counter this with a negative Horizontal Expansion, but that setting is not respected per model and just follows the global model setting.

![image](https://user-images.githubusercontent.com/95894900/151096475-8f0be568-a12e-4c9e-8fb7-c75f545814fe.png)

**Comment by mattwn** [comment]
This might not be limited to Spiralize. Considering a third way to print a vase-like part (Wall Count = 1 and Infill = 0), also fails to comb correctly. 

It should be noted that all three cases print in the same general order, `;TYPE:SUPPORT` first, and `;TYPE:OUTER-WALL` second.

Why does Surface mode comb/avoid properly on the transition from Supports to Wall, but the other methods to not?

![image](https://user-images.githubusercontent.com/95894900/151173208-8305ad07-09df-4293-b3a3-451d4646ba8a.png)

**Comment by mattwn** [comment]
This "feature request" became sprawling, contained multiple bugs, and is not really a feature request. I will close this issue and post the two found bugs individually so they may be tracked.
